Adding OrderTableDatastore and Meta caching integration

prettyboymp opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #45550.

**Background:**
The current OrderCaching works by attempting to cache the Order objects.  However, the current implementation of `WC_Data::__sleep()/__wakeup()` only allow the ID to be serialized.  The reason for this was to avoid serializing the datastore with a data object, but also prevents issues where complex hierarchies of objects can end up getting cached with the order object.  The end result means that any unserialized WC_Data object has to be reinitialized completely, requiring DB queries to retrieve the data for re-hydration.

This PR introduces the following that are disabled by default:

1. Caching of raw order data is added to the HPOS DataStore layer reducing the queries needed to initialize order objects.
2. OrderCache (which is used to cache instantiated WC_Order class objects) now uses an `order_object` cache group that is set as non-persistent. We don't want to save the order objects in persistent storage and the time spent Instantiating a WC_Order object is small, but can add up.  So non-persistent cache is used to avoid re-instantiating the same object over and over again during a single request.
3. Introduces a new `WOOCOMMERCE_ENABLE_ALPHA_FEATURE_TESTING` constant that can be used during initial alpha integration of new features/enhancements.  When defined, the option to enable order data caching is available through the advanced features settings.

Benefits:
1. An entity doesn't need to care where it is stored (e.g. which datastore) or about it's own caching.  So we can remove some of the work-arounds added to deal with cache when syncing.  
2. Caching a set of Orders continues to only cache the set of IDs instead of the entire object.  This ensures that the orders restored from the set are the latest versions of those orders as provided in the database.

**Alternative Considered:**
I looked at improving the serialization handling of WC_Data and it's child classes so that we could serialize all the relevant data with the entities without connected services, like the data stores.  There were a couple issues with this:

First, PHP limitations meant that the serialization/deserialization would need to be pretty complex: PHP doesn't have a simple way to use the default serialization and just exclude certain properties.  To properly restore an object, each class in the hierarchy would need to potentially  deal with any private properties it defines. With extended classes from other plugins, those child classes would need to be properly modified for this to work or compatibility would break.  

Second, there is already a bit of juggling going on around caching orders because the data stores have accessible public API's for making updates.  Shifting the caching to them allows them to directly control the cache and update it when needed.


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Unit tests have been updated to validate proper cache invalidation.

The following can be used to test the functionality of the OrdersTableDataStore caching.  :
1. Create a file named `PR46023Tests.php` somewhere within the `plugins/woocommerce/tests/php/src` diretory.

```php
<?php

/**
 * Test for https://github.com/woocommerce/woocommerce/pull/46023 to verify that the caching actually makes
 * performance improvements over the current cache handling by caching the order data not currently preserved
 * with the WC_Data objects.  It works by forcing cached objects/arrays to be serialized like they would be with a 
 * persistent cache solution. 
 */
class PR46023Tests extends WC_Unit_Test_Case {

	use \Automattic\WooCommerce\RestApi\UnitTests\HPOSToggleTrait;

	public function setUp(): void {
		parent::setUp();
		add_filter( 'wc_allow_changing_orders_storage_while_sync_is_pending', '__return_true' );
		$this->setup_cot();
		$GLOBALS['wp_object_cache'] = new Serializing_Cache_Proxy( $GLOBALS['wp_object_cache'] );
	}

	public function tearDown(): void {
		$this->clean_up_cot_setup();
		remove_filter( 'wc_allow_changing_orders_storage_while_sync_is_pending', '__return_true' );
		parent::tearDown();
	}

	public function test_order_cache_perfmance() {
		global $wpdb;
		$order_1 = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::create_order();
		$order_2 = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::create_order();
		$order_3 = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::create_order();
		$order_4 = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::create_order();
		$order_5 = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::create_order();
		$order_6 = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::create_order();

		wc_get_orders( [
			$order_1->get_id(),
			$order_2->get_id(),
			$order_3->get_id(),
			$order_4->get_id(),
			$order_5->get_id(),
			$order_6->get_id(),
		] );

		$initial_queries = $wpdb->num_queries;
		wc_get_order( $order_1->get_id() );
		wc_get_order( $order_2->get_id() );
		wc_get_order( $order_3->get_id() );
		wc_get_order( $order_4->get_id() );
		wc_get_order( $order_5->get_id() );
		wc_get_order( $order_6->get_id() );

		$final_queries = $wpdb->num_queries;
		$this->assertEquals( 0, $final_queries - $initial_queries, 'Expected no queries to occur retrieving previously retrieved Orders' );
	}
}

/**
 * Proxy class used to mimic objects
 */
class Serializing_Cache_Proxy {
	public $original_cache_instance;

	public function __construct( $original_cache_instance ) {
		$this->original_cache_instance = $original_cache_instance;
	}

	public function __call( $function, $args ) {
		return $this->original_cache_instance->$function( ...$args );
	}

	public function get( $key, $group = 'default', $force = false, &$found = null ) {
		$data = $this->original_cache_instance->get( $key, $group, $found, $found );
		if ( is_object( $data ) || is_array( $data ) ) {
			return unserialize( serialize( $data ) );
		}

		return $data;
	}
}
```
2. Run the test unit tests, filtering to the new test with `--filter=PR46023Tests`

#### Manual Testing

1. Setup an environment with persistent object caching drop-in setup.
2. Under wp-admin/admin.php?page=wc-settings&tab=advanced&section=features, verify that the HPOS Data Caching Feature is not shown.
3. Add the following to the wp-config.php: `define( 'WOOCOMMERCE_ENABLE_ALPHA_FEATURE_TESTING', true );`
4. Reload the advanced->features and verify that the option now shows.
![Screenshot 2024-08-15 at 2 50 55 PM](https://github.com/user-attachments/assets/4d4e041a-cac1-40ec-bde4-fdfc6488a3d7)
5. Disable persistent cache by temporarily removing the drop-in.
6. Reload the advanced->features and verify that the option is visible, but disabled noting that it is only suggested with external object caching.
![Screenshot 2024-08-15 at 2 52 38 PM](https://github.com/user-attachments/assets/175cbaa2-676e-400e-b732-dfdca0a3dd26)
7. Re-enable external object caching.
8. Enable HPOS and the HPOS Data Caching.
9. Place an order through the site
10. Within the admin, validate that the data is as expected.
11. Update the order within the admin, including any meta data.
12. Verify that the data is updated as expected.
13. Revising the order details as a customer and make sure the data reflects the latest data.

#### Additional Automated testing:

1. Modify `OrderUtil:: custom_orders_table_datastore_cache_enabled()` to always return `true`;
2. Run the woocommerce core testsuites.

prettyboymp committed:
Adding initial handling for OrderTableDatastore caching integration

prettyboymp committed:
fix ::cache_objects() mapping usage

prettyboymp committed:
Extend the WPCacheEngineTest tests to cover new multi-set/get methods

prettyboymp committed:
Clear order data cache after ::persist_order_to_db()

prettyboymp committed:
clear child order cache after a direct parent_order_id update

prettyboymp committed:
make sure we clear cache after any direct db writes

prettyboymp committed:
simplify interface to clear object cache

prettyboymp committed:
add changelog

prettyboymp committed:
fixing lint errors

prettyboymp committed:
more lint fixes

prettyboymp committed:
continuing lint fixes

prettyboymp committed:
additional lint fixes

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
fix typo in test

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp commented:
Pausing review on this until I can come up with a better testing process.

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
Adding cache invalidation after OrderTable deletion
Renaming methods to invalidate_ rather than delete, marking as @internal

prettyboymp committed:
Add test to verify cache invalidation after order migration

prettyboymp committed:
lint fixes

prettyboymp committed:
lint fixes

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
update test_get_order_doesnt_return_invalid_cached_order after #46393 merge.

prettyboymp committed:
lint fixes

prettyboymp committed:
fix bug where partially cached set didn't correctly load uncached orders

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

vedanshujain commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -98,13 +100,16 @@ public function migrate_orders( array $order_post_ids ): void {
 
 			$this->handle_migration_error( $order_post_ids, $errors, $exception, $using_transactions, $name );
+			if ( ! $using_transactions ) {
+				wc_get_container()->get( OrdersTableDataStore::class )->invalidate_cache_for_objects( $order_post_ids );
```
Just curious, why is this required? Wouldn't the object be correctly cached at this point? (and same for cache invalidation at line 112)

prettyboymp replied:
The `PostsToOrdersMigrationController`  runs direct queries against the tables/data the  `OrdersTableDatastore` manages without running those queries through it.  So there is nothing in that chain that would clear the object in the datastore's cache.  We need to explicitly invalidate the cache for those objects since we're modifying data outside of the datastore context.  
Ideally, the migration controller would use the same order objects generated from the CPT Data store and sent it too the COT datastores for saving/syncing/migration and we wouldn't have to expose cache invalidation externally.  But I imagine that had to bypassed because of complexities in data flow because we don't have a clear boundary to keep business logic out of the datastore layers so applying the ideal flow would have been more complex.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/CustomMetaDataStore.php:
```diff
@@ -5,13 +5,22 @@
+	 *
+	 * @return string
+	 */
+	abstract protected function get_cache_group();
```
Note that this change won't be backward compatible, maybe we can instead make this function a no-op with a developer warning.

prettyboymp replied:
Good point.  We'll need a good default to make sure it still works correctly. I'm thinking that falling back to the value of `::get_table_name()` may be a good option.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -526,6 +527,57 @@ public function get_all_order_column_mappings() {
+	 * @return bool[] Array of return values, grouped by the object_id. Each value is either true on success, or false
+	 *                if the contents were not deleted.
+	 */
+	public function invalidate_cache_for_objects( array $object_ids ): array {
```
This function should probably be part of the WPCacheEngine class, it seems very generic, and have a very low dependency on the internals of the datastore class.

prettyboymp replied:
The dependency it has on the internals of the datastore class are how objects are cached. Caching is meant to be opaque, so what happens in this method is specific to this implementation, e.g, what cache group/keying to use or if we add a similar implementation on top of a CPT datastore, we would likely call `clean_post_cache()` within this method. 

Ideally, this method wouldn't even need to be public since the fact that there is caching isn't really something that should be known outside of the datastore.  Ideally, we also wouldn't have to call the meta datastore to invalidate its cache for an object. But, because there are existing pathways that directly modify data that the `OrdersTableDataStore` manages, we need a way for those paths to be able to notify the datastore that the data was modified externally.

Would it make more sense to leave this method here if the method were called something like `::objects_modified_externally( array $object-ids )`?

vedanshujain reviewed (commented):
Great work! Submitting some intial feedback

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
changing CustomMetaDataStore::get_cache_group() to non-abstract method for backwards compatibility

prettyboymp committed:
Attempt to retrieve order types from cache prior to querying

prettyboymp commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -3076,6 +3153,7 @@ protected function after_meta_change( &$order, $meta ) {
 		} else {
 			$order_cache = wc_get_container()->get( OrderCache::class );
 			$order_cache->remove( $order->get_id() );
+			$this->invalidate_cache_for_objects( array( $order->get_id() ) );
```
I would like to leave this out as all the callers to `::after_meta_change()` already flush meta cache and there isn't really a direct reason to flush the cache for the object too.  However, having this here is needed for Woo Subscriptions compatibility when `$subscription->update_dates()` is used to update the created date. This may be a trivial incompatibility, though.

Removing this line takes off another 10 queries from the checkout flow.

prettyboymp commented:
@vedanshujain, I'm not sure the status of this, but please hold off on merging it at this time.  While it makes improvements to checkout performance, I'm finding that it slows down some API requests.  It seems that with the OrderCache still in place and the Order::constructor() getting called during `::__wakeup()`, the splitting of cache requests ends up performance worse than if nothing is cached at all.  
I'm still digging into why that is.

At the moment, i think OrderCache may still be helpful as an in-memory runtime cache that doesn't push back to persistent cache but I need more time to explore it.

prettyboymp commented:
> @vedanshujain, I'm not sure the status of this, but please hold off on merging it at this time. While it makes improvements to checkout performance, I'm finding that it slows down some API requests. It seems that with the OrderCache still in place and the Order::constructor() getting called during `::__wakeup()`, the splitting of cache requests ends up performance worse than if nothing is cached at all. I'm still digging into why that is.
> 
> At the moment, i think OrderCache may still be helpful as an in-memory runtime cache that doesn't push back to persistent cache but I need more time to explore it.

Findings:
We found that memcache was extremely slow at retrieving the order data when the Order class object was also cached.  In this case the Order object would be retrieved from cache and start deserialization which would then callback to load the order data from cache.  I believe this nested cache call was why it was so slow, but haven't tested separately to confirm.

Because of this, the caching performs much better with the OrderCache (objects) storing a copy of the Order object in-memory for a single request and not sending them to persistent cache.   There multiple ways to do this, however.  Specifically for the memcache plugin, you can set the `orders` group as a `no_mc_groups` to prevent it from being a persistent group.  This works best in that it allows any existing work arounds for invalidating Order cache by directly clearing wp_object_cache to still work.  However, this only works for the memcache plugin.  We don't know if this recursive cache get call effects other caching engines.

prettyboymp commented:
One other finding, which is somewhat related to this but already an issue that just doesn't necessarily appear without this caching present:

Instantiating an order directly via `new WC_Order( $order_id )` will load custom order types, e.g. subscriptions, with the OrdersTableDataStore instead of the one for the Subscription.  This causes some warnings if the column mappings are different between the two.

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
Put Order data caching behind a feature
Switch OrderCache to use request level caching for better performance than the current persisten implementation

prettyboymp commented:
I initially attempted to create an implementation of order data caching that could sit outside of core.  However, the options to implement it were just too fragile and would likely break with any number of plugins or changes to core because of the tight coupling/dependency between order types and their datastore implementation. 

Instead, in the latest push, I've implemented the data caching behind an experimental feature flag.  I also modified OrderCache to run in-memory only instead of pushing to persistent cache.  The current implementation really slows things down.  

These two things should probably get merged separately.  I'd like to work on the OrderCcache change getting included by itself under it's own merit of performance improvements separate to the optional data caching feature.

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
fix syntax errors after find/replace

prettyboymp committed:
Remove DEFAULT group from InMemoryCache

prettyboymp committed:
lint fixes

prettyboymp committed:
lint fixes

prettyboymp committed:
lint fixes

prettyboymp committed:
more lint fixes
trigger cache flush when caching feature is initially turned on.

prettyboymp committed:
more lint fixes

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
Share the RequestLevelCache

prettyboymp committed:
fix call to parent

prettyboymp committed:
Remove test test_get_order_doesnt_return_invalid_cached_order as there moving OrderCache to be non-persistent removes the possible race condition between cache and DB propogation

prettyboymp committed:
fix bug in calling parent::get_meta_data_for_object_ids()

prettyboymp committed:
Flush the OrderCache directly instead of flushing all cache to simulate a new request when calling reset_order_data_store_state

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
cleanup feature implementation, fix some bugs around callable checks

prettyboymp committed:
Introduce WOOCOMMERCE_ENABLE_ALPHA_FEATURE_TESTING constant, avoid displaying the order data caching option unless the contant is true

prettyboymp committed:
Removing InMemoryOrderCache in favor of making the `orders` cache group non-persistent for backward compatibility.

prettyboymp committed:
Add docblock to maybe_set_order_cache_group_as_non_persistent

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
be sure to instantiate $order_data

prettyboymp committed:
Move order cache to a different cache group.

The orders cache group gets used by a lot of things, including resource heavy and slow reporting.  So there is a penalty for making that entire group non-persistent.  To work around this, we'll move order object cache to it's own group and make it non-persistent.

prettyboymp committed:
Go ahead and cache the order data from a call to `::get_orders_type()`

Most calls to `::get_orders_type()` end up requiring the entire object later in the request since it's mostly used to determine what object class to load the order object with.  Instead of running an extra query, this will go ahead and load all the objedct data assuming the entire object will be loaded soon enough.

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
fix lint issue

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

kalessil reviewed (commented):
@prettyboymp: thank you for the detailed PR description - it greatly helped!

My simplified understanding is that we are introducing in-memory caching.

Also, my following feedback assumes that I understood the concept correctly:
- Well done with iterative rollout (as alfa-feature) and using WP-cache for in-memory storage!
- I understand that we are yet tinkering here. Still, the complexity of the changes is high, and I suggest refactoring things a bit before rolling out the feature as alfa: the fast route would be to aggregate/extend the classes heavily affected by caching mechanisms (where possible) and place new bits in there.

Do you think we could do something about this?

prettyboymp committed:
fix comment and indentation issue.

prettyboymp committed:
simplify ovveride methods in OrdersTableDataStoreMeta now that the implementation is stable

prettyboymp committed:
Some refactoring in data stores to break out cache and database interactions into their own methods.

Breaking these out should simplify future refactoring to possibly move DB and Cache layer interactions into their own classes and let the datastore delegate to those.

prettyboymp committed:
Fix lint issues

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp commented:
> * I understand that we are yet tinkering here. Still, the complexity of the changes is high, and I suggest refactoring things a bit before rolling out the feature as alfa: the fast route would be to aggregate/extend the classes heavily affected by caching mechanisms (where possible) and place new bits in there.
> 
> Do you think we could do something about this?

@kalessil, I pushed some refactoring around the datastore methods that were both accessing cache and the db.  I broke these out into db specific methods and cache specific methods and left the merging/management of the data in the original method.  Think this simplifies those and better defines where the abstraction/separation should go for future iterations.

Because the current abstractions of the DB layer within the datastores uses inheritance rather than composition, the future iterations won't be easy to implement without further compatibility concerns and likely temporary additional complexity, which is why I'd like to address those later.

kalessil commented on plugins/woocommerce/src/Internal/DataStores/CustomMetaDataStore.php:
```diff
@@ -53,31 +53,20 @@ protected function get_db_info() {
-			)
-		);
-		// phpcs:enable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$raw_meta_data = $this->get_meta_data_for_object_ids( array( $object->get_id() ) );
```
Nit: `$object->get_id()` can be extracted into a local variable to reduce the number of the method invocations.

kalessil commented on plugins/woocommerce/src/Internal/DataStores/CustomMetaDataStore.php:
```diff
@@ -265,4 +262,47 @@ public function get_meta_keys( $limit = 100, $order = 'ASC', $include_private =
+		// phpcs:enable
+
+		foreach ( $meta_rows as $meta_row ) {
+			if ( ! isset( $meta_data[ $meta_row->object_id ] ) ) {
```
Nit: we should probably array fill instead before the loop - that'd reduce cognitive load abit.

prettyboymp replied:
Thanks for this.  After re-reviewing this, filling the array ahead of time for the $object_ids works out better programmatically here as well.  This made me notice that, though unlikely, an order with no meta data would result in no value set for the object_id.  This would both mean it would always be a cache miss and cause some warnings in `OrdersTableDataStore::init_order_record()`.

kalessil commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -529,6 +531,71 @@ public function get_all_order_column_mappings() {
+		$return_values = array();
+
+		foreach ( $order_ids as $order_id ) {
+			$return_values[ $order_id ] = $cache_engine->delete_cached_object( $order_id, $this->get_cache_group() );
```
Nit: `$this->get_cache_group()` can be moved into a local variable before the loop to reduce the number of the method invocations.

kalessil commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -529,6 +531,71 @@ public function get_all_order_column_mappings() {
+			$meta_invalidated = $this->data_store_meta->clear_all_cached_data();
+		}
+
+		return $orders_invalidated && $meta_invalidated;
```
Nit: just from the workflow here, `$meta_invalidated` might be not defined.

prettyboymp replied:
Good catch.

kalessil commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -1645,26 +1726,52 @@ private function set_order_prop( \WC_Abstract_Order $order, string $prop_name, $
 	 */
-	private function get_order_data_for_id( $id ) {
-		$results = $this->get_order_data_for_ids( array( $id ) );
+	protected function get_order_data_for_ids( $ids ): array {
```
Can we declare `$ids` as `array $ids` here?

prettyboymp replied:
I was hesitant to do this because `OrdersTableDataStore::get_orders_type()` doesn't strictly declare the argument type.  I made a change to that caller to temporarily handle this potential compatibility issue in hopes we can eventually push strict typing up the call stack.

kalessil reviewed (commented):
@prettyboymp: with the latest updates, it's much easier to grasp, thanks for the updates!

Also leaving some implementation-related hints, feel free to ignore them or let me know if I should turn geeky mode and do deeper implementation review.

prettyboymp committed:
Address feedback.

prettyboymp committed:
fix lint issues

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
fix undefined $order_data variable

prettyboymp committed:
fix docblock parameter declarations

kalessil reviewed (approved):
LGTM, thanks for updating the PR!

@prettyboymp, what'd be our next steps to push the caching forward?

coreymckrill commented on plugins/woocommerce/src/Caches/OrderCache.php:
```diff
@@ -2,21 +2,29 @@
+			 * The use of datastore caching moves persistent data caching to the datastore. Order object caching then only
+			 * acts as request level caching as the `order_objects` cache group is set as non-persistent.
+			 */
+			return 'order_objects';
```
Not sure if this is a valid concern or not, but adding it here just in case. It appears that this `object_type` value is used for [more than just a cache key](https://github.com/woocommerce/woocommerce/blob/038e0e910fa415c386f37da04c38e9d61fff3039/plugins/woocommerce/src/Caching/ObjectCache.php#L60-L63). In particular, it's also used for hook names. It seems like there could be backcompat issues here if we conditionally change this. But I'm not sure what the alternative would be.

prettyboymp replied:
I'm going to update this docblock to remove the reference to hooks/options. The only built in `CacheEngine` implementation is `WPCacheEngine.php` which doesn't have any underlying hooks or option usage.  

I think the compatibility concern is limited.  It would require that an extension both overwrote the cache engine implementation in a way where it depended on the object type returned from the default cache implementation for some behavior but hard coded the previous default in hook usage, or vice versa.

coreymckrill commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStoreMeta.php:
```diff
@@ -1,17 +1,30 @@
 /**
  * OrdersTableDataStoreMeta class file.
+ *
+ * @phpcs:disable Universal.NamingConventions.NoReservedKeywordParameterNames.objectFound
```
This should be re-enabled sometime after the offending code lines.

prettyboymp committed:
Remove reference to option/hook values for docblock of ObjectCache::get_object_type()
Adjust phpcs enabling/disabling of NoReservedKeywordParameterNames in OrdersTableDataStoreMeta

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer

prettyboymp committed:
Fix merge conflicts and change to public filter callbacks

prettyboymp committed:
lint fix

prettyboymp committed:
Merge branch 'trunk' into fix/45550-cache-orders-at-datastore-layer
